### PR TITLE
#272 Mod Support

### LINKED ISSUE #272: [Enhancement] MODS environment variable
Acceptance Criteria:
* `installmod` odin command.
* Ability to detect if the mod is a `dll` or valheim plus.
* staging the mod information.
* installed mods configuration file.

### PR DESCRIPTION
**Pull request by mbround18**
# Description

## Contributions
- Added mod downloader
- Added type
- Closes #272
- Refactored rust main to be simpler.
- Added preset env for cron jobs to carry over cron data. 
- Moved to directory variables. 

## Checklist

- [x] I added one or multiple labels which best describes this PR.
- [x] I have tested the changes locally.
- [x] This PR has a reviewer on it. 
- [x] I have validated my changes in a docker container and on Ubuntu. (Only needed for Odin or Docker Changes)

### CODE REVIEW

**Comment by LovecraftianHorror** [comment]
Still sifting through all the changes. I should have a review finished up tomorrow!

File: src/mods/mod.rs
Comment (LovecraftianHorror) [inline src/mods/mod.rs:49]: Shouldn't this also be a hard error?

File: src/scripts/start_valheim.sh
Comment (LovecraftianHorror) [inline src/scripts/start_valheim.sh:104]: Can we add in a warning or error if `$TYPE` is vanilla and `$MODS` is set to something that the mods won't actually be installed without one of the supported mod frameworks?

File: src/scripts/start_valheim.sh
Comment (LovecraftianHorror) [inline src/scripts/start_valheim.sh:67]: It may be somewhere and I just missed it, but it looks like there's never a check for a valid `$TYPE` so someone could accidentally misspell the name of the mod framework and end up in a vanilla world on accident.

File: src/mods/mod.rs
Comment (LovecraftianHorror) [inline src/mods/mod.rs:128]: Another place where it might be appropriate to have a hard error

**Review by LovecraftianHorror — CHANGES_REQUESTED** [review]
Alright that's my first pass over everything. I don't want to keep holding this up since I'm still trying to get a grasp on everything, but I think it's onto manual testing for me which will take a while so I'll go ahead and submit this. There's nothing deal-breaking in here just some stuff for clarity and cleaning up. I think the big things I want settled are:

1. How to treat mod installation failures? Currently there are several points where the failure silently goes by (usually with a log but without an exit status or other hard error). This means that someone could think that the container launched with all the mods working when really something failed down the line
2. It currently looks like it tries to install the mod framework and mods on every launch if it's set and `$UPDATE_ON_STARTUP` is set, but nothing is done inside the `auto_update` cron job. The only problems there are it's hard to know if there was an actual update or not, but shutting down the container and restarting would still be required. I'd be fine not dealing with this quite yet since there's already a lot going on inside this PR, but we might want to keep track of some metadata with the downloads (filehashes would probably be the best here) and use that to check if it seems like anything is actually updated).
3. I didn't dig into this too far since there was a lot to go through, but how is downloading and extracting the same filename multiple times handled? Will it overwrite the previous contents (which is likely the desired behavior)? I know that copying will, I just want to make sure all the previous steps do too.
4. We don't have to do it now, but eventually cleaning up all the old `zip` files and originally extracted contents would be good to do, if it's not already handled.

And just to add an overview of the process since it took a bit of digging for me to understand.

## PR Overview

This PR adds the ability to use different mod frameworks as well as downloading mods based on a provided URL. This is exposed through a `$TYPE` env var which specifies either vanilla or a tag indicating the desired mod framework along with the `$MODS` env var which is a list of mod urls separated by a comma and new-line.

### Outside `odin`

Externally this will check for the presence of a mod framework specified in `$TYPE` and follow specific install procedures for the specified mod framework. Following the framework installation the container will also go through each url within `$MODS` and install the mod using `odin`'s new installmod subcommand.

### Inside `odin`

`odin installmod URL` attempts to download the file pointed to by `URL` into a staging location. The filename will be the filename from the `URL` if detected and falls back to `{{ md5_hash(URL) }}.zip` if no filename is detected.

From there the zip file is extracted to the staging location as well and will have the contents copied over (there is some nuance on whether or not there is a `manifest.json` that would be good to manually test both cases).

File: README.md
Comment (mbround18) [reply README.md:53]: Probably, I can put together some good examples for people to hopefully grasp the syntax from.

File: src/commands/install_mod.rs
Comment (mbround18) [reply src/commands/install_mod.rs:13]: I figure for a mod, a silent fail is appropriate because in some cases they might already have the mod installed and the url is purely stale. What I could add is it exiting with code 1 on a failure and than in the shell script trapping that failure with a more noticeable message. What are your thoughts?

File: src/cli.yaml
Comment (mbround18) [reply src/cli.yaml:123]: does clap support nested subcommands? 0.o

File: src/mods/bepinex.rs
Comment (mbround18) [reply src/mods/bepinex.rs:26]: Ooo, I did now know this existed! I can make the change :)

File: src/mods/mod.rs
Comment (mbround18) [reply src/mods/mod.rs:67]: I agree it should be a hard error, I can make the change :)

File: src/mods/mod.rs
Comment (mbround18) [reply src/mods/mod.rs:97]: Not in this case because String != &String

Comment (LovecraftianHorror) [reply src/mods/mod.rs:97]: There isn't any need for the ref though

File: src/scripts/start_valheim.sh
Comment (mbround18) [reply src/scripts/start_valheim.sh:115]: The IFS should take care of that, Are you talking about url sanitization?

File: src/scripts/start_valheim.sh
Comment (mbround18) [reply src/scripts/start_valheim.sh:104]: Yee

File: src/scripts/start_valheim.sh
Comment (mbround18) [reply src/scripts/start_valheim.sh:67]: Expected, KIS approach :P

File: src/scripts/start_valheim.sh
Comment (LovecraftianHorror) [reply src/scripts/start_valheim.sh:115]: Oh that is true, I didn't know IFS was so magic, and it looks like the `$` is needed? Or at least I had weird issues without it when `n`s were in the URL

File: src/scripts/start_valheim.sh
Comment (LovecraftianHorror) [reply src/scripts/start_valheim.sh:106]: Just kidding it is required

File: src/mods/mod.rs
Comment (mbround18) [reply src/mods/mod.rs:103]: By ref do you mean the barrow char of '&' ? 0.o

File: src/scripts/start_valheim.sh
Comment (mbround18) [reply src/scripts/start_valheim.sh]: yee it is

File: README.md
Comment (LovecraftianHorror) [inline README.md:52]: There's also bepinexfull as an option too

File: Cargo.toml
Comment (LovecraftianHorror) [inline Cargo.toml:25]: Should have checked before, but `glob` isn't used anymore

File: src/commands/install_mod.rs
Comment (LovecraftianHorror) [reply src/commands/install_mod.rs:13]: I think that failing within odin and then ignoring that for the shell script would be a good idea! I'll leave the final decision up to you though

Comment (LovecraftianHorror) [reply src/commands/install_mod.rs:13]: It looks like all the errors raised by this are from the staging directory and an invalid mod url which seems fair to hard error on both of those, and then it can just be ignored by the container

**Review by LovecraftianHorror — CHANGES_REQUESTED** [review]
Looks good, only a couple small things left.

I haven't manually tested, but it seems like you're going over things pretty well already, so I'm not that worried on that front

**Review by LovecraftianHorror — APPROVED** [review]
All of the changes look good to me! I'll leave the final testing up to you since you're much more familiar with dealing with Valheim mods than me.

I do think it's worth noting that the current behavior means that the program will try to access whatever file is downloaded as a zip archive if it's not a `dll` or `cfg` file. It is specified that `dll`, `cfg`, and `zip` are the only supported formats, but I just figured it's worth noting since this would cause `odin` to panic when it hits some other format. The alternative would be to check for a valid file type again after following the redirect and then exit with an error message if the file type doesn't match one of our allowed options.

**Pull request merged**
